File: app/src/main/java/ua/hospes/nfs/marathon/domain/race/models/RaceItem.java
package ua.hospes.nfs.marathon.domain.race.models;

import ua.hospes.nfs.marathon.domain.sessions.models.Session;
import ua.hospes.nfs.marathon.domain.team.models.Team;

/**
 * @author Andrew Khloponin
 */
public class RaceItem {
    private final int id;
    private final Team team;
    private int teamNumber = -1;
    private Session session;

    private RaceItemDetails details;


    public RaceItem(Team team) {
        this.id = -1;
        this.team = team;
    }

    public RaceItem(int id, Team team) {
        this.id = id;
        this.team = team;
    }


    //region Getters
    public int getId() {
        return id;
    }

    public Team getTeam() {
        return team;
    }

    public Session getSession() {
        return session;
    }

    public int getTeamNumber() {
        return teamNumber;
    }

    public RaceItemDetails getDetails() {
        return details;
    }
    //endregion

    //region Setters
    public void setSession(Session session) {
        this.session = session;
    }

    public void setTeamNumber(int teamNumber) {
        this.teamNumber = teamNumber;
    }

    public void setDetails(RaceItemDetails details) {
        this.details = details;
    }
    //endregion


    @Override
    public String toString() {
        return "RaceItem{" +
                "id=" + id +
                ", team=" + team +
                ", teamNumber=" + teamNumber +
                ", session=" + session +
                ", details=" + details +
                '}';
    }
}

File: app/src/main/java/ua/hospes/rtm/domain/race/models/RaceItem.kt
package ua.hospes.rtm.domain.race.models

import ua.hospes.rtm.data.model.RaceDto
import ua.hospes.rtm.domain.sessions.Session
import ua.hospes.rtm.domain.sessions.toDomain
import ua.hospes.rtm.domain.sessions.toDto
import ua.hospes.rtm.domain.team.Team
import ua.hospes.rtm.domain.team.toDomain
import ua.hospes.rtm.domain.team.toDto

data class RaceItem(
    val id: Long = 0,
    val teamNumber: Int,
    val team: Team,
    val session: Session? = null,
    val details: RaceItemDetails? = null
)

internal fun RaceItem.toDto(): RaceDto = RaceDto(
    id = id,
    team = team.toDto(),
    teamNumber = teamNumber,
    session = session?.toDto(),
)

internal fun RaceDto.toDomain(): RaceItem = RaceItem(
    id = id,
    teamNumber = teamNumber,
    team = team.toDomain(),
    session = session?.toDomain(),
    details = details?.let {
        RaceItemDetails(pitStops = it.pitStops, completedDriversDuration = it.completedDriversDuration.clone())
    },
)
